rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,865 @@
+            it's the output size of the underlying `RasaFeatureCombiningLayer` if the
+            `Ffnn` block has 0 layers. `seq_length` is the sum of the sequence dimension
+            sizes of sequence- and sentence-level features (for details, see the output
+            shape of `RasaFeatureCombiningLayer`). If both feature types are present,
+            then `seq_length` will be 1 + the length of the longest sequence of real
+            tokens across all examples in the given batch.
+        seq_sent_features: `(batch_size, seq_length, hidden_dim)`, where `hidden_dim` is
+            the output size of the underlying `Ffnn` block, or the output size of the
+            underlying `RasaFeatureCombiningLayer` if the `Ffnn` block has 0 layers.
+        mask_combined_sequence_sentence: `(batch_size, seq_length, hidden_dim)`
+        token_ids: `(batch_size, seq_length, id_dim)` where id_dim is unimportant and
+            it's the last-dimension size of the first sequence-level dense feature
+            if any is present, and 2 otherwise. Empty tensor if not doing MLM.
+        mlm_boolean_mask: `(batch_size, seq_length, 1)`, empty tensor if not doing MLM.
+        attention_weights: `(transformer_layers, batch_size, num_transformer_heads,
+            seq_length, seq_length)`, empty tensor if the transformer has 0 layers.
+
+    Raises:
+        A RasaException if no feature signatures for sequence-level features are
+            provided.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        attribute_signature: Dict[Text, List[FeatureSignature]],
+        config: Dict[Text, Any],
+    ) -> None:
+        if not attribute_signature or not attribute_signature.get(SEQUENCE, []):
+            raise RasaException(
+                "The attribute signature must contain some sequence-level feature\
+                signatures but none were found."
+            )
+
+        super().__init__(name=f"rasa_sequence_layer_{attribute}")
+
+        self._tf_layers: Dict[Text, Any] = {
+            "feature_combining_layer": RasaFeatureCombiningLayer(
+                attribute, attribute_signature, config
+            ),
+            "ffnn": layers.Ffnn(
+                config[HIDDEN_LAYERS_SIZES][attribute],
+                config[DROP_RATE],
+                config[REGULARIZATION_CONSTANT],
+                config[WEIGHT_SPARSITY],
+                layer_name_suffix=attribute,
+            ),
+        }
+
+        self._prepare_masked_language_modeling(attribute, attribute_signature, config)
+
+        transformer_layers, transformer_units = self._get_transformer_dimensions(
+            attribute, config
+        )
+        self._prepare_transformer(
+            attribute, config, transformer_layers, transformer_units
+        )
+
+        self.output_units = self._calculate_output_units(
+            attribute, transformer_layers, transformer_units, config
+        )
+
+    def _prepare_transformer(
```

Comment:
Can't we merge this function with `_prepare_transformer_layer` function?

Reply:
It makes sense to me too @dakshvar22 -- to make the function global. Then the only thing that needs to change is setting an entry in the layers dictionary. Would you say that just returning the prepared layer is preferred, like this:
```python
transformer = prepare_transformer(config, etc)
self._tf_layers["transformer"] = transformer
```

Reply:
why can't we directly set it, like this - 
`self._tf_layers["transformer"] = prepare_transformer(config, etc)`

If `prepare_transformer` is returning multiple return values then you can escape them with `_`.

Reply:
Oh, yeah, sure, I meant exactly this, not sure why I expanded it over 2 lines 🙂